File: code_snippets/java/src/main/java/get_started/Greeter.java
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate examples,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/
 */

package my.example;

import dev.restate.sdk.Context;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.Service;
import dev.restate.sdk.http.vertx.RestateHttpEndpointBuilder;

import java.time.Duration;

/**
 * Template of a Restate service and handler
 * Have a look at the Java QuickStart to learn how to run this: https://docs.restate.dev/get_started/quickstart?sdk=java
 */
@Service
public class Greeter {

    // <start_here>
    @Handler
    public String greet(Context ctx, String greeting) {

        ctx.run(() -> doWork(1));
        ctx.sleep(Duration.ofSeconds(2));
        ctx.run(() -> doWork(2));
        ctx.sleep(Duration.ofSeconds(2));
        ctx.run(() -> doWork(3));
        ctx.sleep(Duration.ofSeconds(2));
        ctx.run(() -> doWork(4));

        return greeting;
    }

    private void doWork(int taskNb){
        System.out.println("Task " + taskNb +
                " started at " + System.currentTimeMillis());
    }
    // <end_here>

    public static void main(String[] args) {
        RestateHttpEndpointBuilder.builder()
                .bind(new Greeter())
                .buildAndListen();
    }
}

File: code_snippets/ts/src/get_started/greeter.ts
// <start_here>
import * as restate from "@restatedev/restate-sdk";
import {Context, service} from "@restatedev/restate-sdk";

const greeter = service({
    name: "greeter",
    handlers: {
        greet: async (ctx: Context, greeting: string) => {
            const message = `${greeting}!`;

            await ctx.run(() => sendNotification(message));
            await ctx.sleep(1000)
            await ctx.run(() => sendReminder(message))

            return message;
        },
    },
});
// <end_here>

restate.endpoint().bind(greeter).listen();

// Simulated notification sending operation
const sendNotification = async (message: string) => {
    if (Math.random() < 0.3) { // 30% chance of failure
        throw new Error("Failed to send notification");
    }
    console.log(`Notification sent: ${message}`);
};


// Simulated reminder sending operation
const sendReminder = async (message: string) => {
    if (Math.random() < 0.3) { // 30% chance of failure
        throw new Error("Failed to send notification");
    }
    console.log(`Reminder sent: ${message}`);
};

